### PR TITLE
Fix the `outline_lcs`

### PR DESCRIPTION
**Pull request by weikengchen**
After careful debugging, it appears that the current implementation of `outline_lcs` has two serious problems:

- It does not correctly create additional constraints. All the additional constraints should be `LC * 1 = V` where V is the new witness variable. In the end, all of them are `V * 1 = V`.

- It does not increase the count of `num_linear_combinations` correctly. So that in Marlin, when the next step wants to pad the matrices to be square, the new linear combinations go to the wrong linear combination entry, causing the outer check to fail.

This PR fixes these two problems and will close this issue in Marlin: https://github.com/arkworks-rs/marlin/issues/56

### CODE REVIEW

**Comment by weikengchen** [comment]
It turns out that I cannot assign a reviewer here. So, let me here cc Pratyush for a review. @Pratyush 

We may, in the future, has an improvement plan for `outline_lcs`, since the resulting saving is not as large as expected.

**Comment by Pratyush** [comment]
Thanks for the PR! Given that there was a subtle bug in the logic here, would it be possible to add some comments documenting the algorithm?

**Comment by weikengchen** [comment]
No problem. On the way!

**Comment by weikengchen** [comment]
The documentation has been added. Please take a look :)

File: relations/src/r1cs/constraint_system.rs
Comment (Pratyush) [inline relations/src/r1cs/constraint_system.rs:429]: Why not use `self.new_witness` here? It would avoid errors from manual management of witness variables.

File: relations/src/r1cs/constraint_system.rs
Comment (Pratyush) [inline relations/src/r1cs/constraint_system.rs]:
```suggestion
                // Add a new constraint for this new witness
                new_witness_linear_combinations.push(outlined_lc.clone());
                new_witness_indices.push(witness_assignment);
```

Also, why not use `self.constrain` here?

**Comment by weikengchen** [comment]
For the `new_witness`, yes, good point, let me use `new_witness` instead.

For the new constraint via `enforce_constraint`, it has a problem---our code is modifying the lc_map but `enforce_constraint` would also push to it, causing a conflict (that our code may overwrite the change). But let me take a more careful look to see I can still use `enforce_constraint` in a black-box way if I save the `lc_map` earlier.

**Comment by Pratyush** [comment]
For the `new_constraint` issue, maybe we can extract the common logic into a separate method that is called in both places?

**Comment by weikengchen** [comment]
Yes. But I feel that probably I can call `new_constraint` directly if I do `self.lc_map = outlined_lcs;` earlier, since these new constraints all use outlined LCs (so no more processing is needed).

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs]:
```suggestion
    /// Count the number of times each LC is used within other LCs in the constraint system
```

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs:246]:
```suggestion
        // Iterate over every lc in constraint system
        for (index, lc) in self.lc_map.iter() {
```

**Comment by weikengchen** [comment]
I use `new_constraint` directly, so the code is now slightly simplified.

I didn't use `new_witness` for a specific reason:
- The `new_witness` somehow has to be in the `for` loop of `self.lc_map`. Moving it out could lead to potential missing assignments.
- But `new_witness` takes `&mut self` as input, but this is forbidden (since `self` is currently used for iteration).
- If I want to allow `new_witness` to use `&mut self`, I need to clone `self.lc_map`, but the memory usage might be too high.

Thus, given that the logic of `new_witness` is simple, I just write the same code again. 

Note: I need to test the upstream Marlin for this PR to see if it is working. So don't merge now!

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [reply relations/src/r1cs/constraint_system.rs]: I know this wasn't part of this PR, but since this was being edited figured might as well improve the comments

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs]: Can a comment be added here as to why sinks are being counted? Its not clear to me atm.

(Also the word sink is non-intuitive to me in this context, since we're not talking about a state machine / probabilistic process of any sort)

(We can make this into a new issue / PR though)

**Comment by Pratyush** [comment]
Hmm in that case we should document that this is the same logic as `new_witness`.

Also, speaking of memory requirements, we should "garbage-collect" the LCs that are inlined/put in new variables, like we do in `inline_all_lcs`

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs]: Can we have a comment to describe what this for loop is doing? It looks `outlined_lcs` is acting as a map to store all prior replacements, and inlining all symbols.

Though I don't get why the core loop works though. (e.g. why does outlined_lcs always contain lc_index in the line `let lc = outlined_lcs.get(&lc_index).expect("should be inlined");`)

**Comment by weikengchen** [comment]
Got it. Let me have some tries.

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs:423]: Can we add a comment explaining the cost. (Namely that we change the arity from `this_len` repeated `this_used_times` to a constraint for `this_len * 1 = {new variable}` and then all of the `this_used_times` entries only have one usage.
```suggestion
            // Cost with no outlining = `lc_len * number of usages`
            // Cost with outlining is one constraint for `(lc_len) * 1 = {new variable}` and 
            // using that single new variable in each of the prior usages.
            // This has total cost `number_of_usages + lc_len + 2`
            if this_used_times * this_len > this_used_times + 2 + this_len {
```

**Review by ValarDragon — COMMENTED** [review]
Added a couple of places for additional comments. None of these should block the PR though

**Comment by ValarDragon** [comment]
Thanks for fixing this bug!

**Comment by weikengchen** [comment]
More review is welcome. Let me merge the suggestions!

File: relations/src/r1cs/constraint_system.rs
Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Resolved by adding more explanations.

**Comment by weikengchen** [comment]
I updated the suggestions. Currently testing with Marlin. 

@ValarDragon can do a final pass over the new comments that try to explain the loop.

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [reply relations/src/r1cs/constraint_system.rs]: Thanks! The algorithm now makes sense to me =)

File: relations/src/r1cs/constraint_system.rs
Comment (Pratyush) [inline relations/src/r1cs/constraint_system.rs]: msp -> map? (Here and elsewhere?)

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Ah! Working on it.

I need to be careful. Recently my typo rate is very high.

**Comment by weikengchen** [comment]
The upstream test passes, and it is ready to merge if no other issues pop up.

**Comment by weikengchen** [comment]
(ps I need someone who has write access.)

**Comment by Pratyush** [comment]
Doing a final review pass, and then I'll merge!

File: relations/src/r1cs/constraint_system.rs
Comment (Pratyush) [inline relations/src/r1cs/constraint_system.rs]: Hmm doesn't this delete LCs when they're made into a new variable? For inlining it doesn't matter, because they're substituted directly into the actual constraints. But here, if the LC is "outlined", we cannot delete it from `new_lc_map`, no?

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Actually, we can. 

If an LC is "outlined", then its entry in `new_lc_map` is now `(1, the new witness variable)`. 

When it is not used anymore (aka, triggering this code), it means that all the subsequent LCs that use this LC has inlined this LC, by extending a term `(1 * coeff, the new witness variable)` into themselves. So, it can be safely deleted from the `new_lc_map`.

And this does not affect the code at about line 411, which creates additional constraints. This is because we copy the "inlined" LC data (not its index) in `new_witness_linear_combinations`.

Comment (Pratyush) [reply relations/src/r1cs/constraint_system.rs]: Ah I misunderstood; this works because the witness *variable* exists, and a corresponding constraint, and so the LC is useless. cool!

In that case, we should definitely extract this common logic between `inline_all_lcs` and `outline_lcs` (by taking in `new_lc_maps`), and leave the above explanation as a comment above the use in `outline_lcs`.

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Like this? Or more than that?
```
// All the LCs depending on `lc_index` have inlined it.
//
// If `lc_index` is outlined, then its witness variable has 
// replaced the original LC.
//
// This lc is not used any more, so remove it.
```

Comment (Pratyush) [reply relations/src/r1cs/constraint_system.rs]: How about something like this?

```rust 
// Deletion is safe for both outlining and inlining:
// * Inlining: the LC is substituted directly into all use 
//   sites, and so once it is fully inlined, it is redundant.
// * Outlining: the LC is associated with a new variable `w`,
//   and a new constraint of the form `lc_data * 1 = w`, where
//   `lc_data` is the actual data in the linear combination.
//   Furthermore, we replace its entry in `new_lc_map` with `(1, w)`.
//   Once `w` is fully inlined, then we can delete the entry from `new_lc_map`
```

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Needs a period at the end of the comments. 

Add this comment to BOTH `inline_all_lcs` and `outline_lcs`, or?

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: And it makes sense!

Comment (Pratyush) [reply relations/src/r1cs/constraint_system.rs]: Let's extract the common logic into a method, and then call this method from `inline_all_lcs` and `outline_lcs`?

Comment (weikengchen) [reply relations/src/r1cs/constraint_system.rs]: Let me have a try on the common logic. Very likely, `inline_all_lcs` and `outline_lcs` will provide a stateful function to this common interface. The common interface asks this function to rewrite each LC, and this function can keep a state.

Comment (Pratyush) [reply relations/src/r1cs/constraint_system.rs]: Yup!

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs]:
```suggestion
            // Inline the LC, unwrapping symbolic LCs that may constitute it,
            // and updating them according to outlinings in prior iterations.
```

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [inline relations/src/r1cs/constraint_system.rs]:
```suggestion
        // `new_lc_map` stores the modified linear combinations.
        //
        // This loop goes through all the LCs in the map, starting from
        // the early ones, and decides whether or not to dedicate a witness
        // variable for this LC.
        //
        // If true, the LC is replaced with 1 * this witness variable.
        // Otherwise, the LC is inlined.
        //
        // Each iteration first updates the LC according to outlinings in prior iterations,
        // and then sees if it should be outlined, and if so adds the outlining to the map.
        //
```

File: relations/src/r1cs/constraint_system.rs
Comment (ValarDragon) [reply relations/src/r1cs/constraint_system.rs]: This doesn't matter for soundness right? Its just for efficiency?

**Review by ValarDragon — APPROVED** [review]
LGTM! Left a couple updates to the comments that helped me understand the structure of the loop. Thanks for fixing the bugs, and significantly improving the documentation around this 

In a future PR, I think we should be able to improve the efficiency by changing `new_lc_map` to being just a map of the differences (`SymbolicLC -> Var`). (Instead of making a copy of all the LC's)

File: relations/src/r1cs/constraint_system.rs
Comment (Pratyush) [reply relations/src/r1cs/constraint_system.rs]: Yes, for memory consumption.

**Comment by weikengchen** [comment]
Thanks for the review. I will resolve the comments soon and also merge the logic of the two functions for fewer repetitions.

(I might have some delays due to sending the Mac for fixing---my "e" key is now an "ee" key.)

**Comment by weikengchen** [comment]
Response to this:

> In a future PR, I think we should be able to improve the efficiency by changing new_lc_map to being just a map of the differences (SymbolicLC -> Var). (Instead of making a copy of all the LC's)

Indeed almost all lcs would be rewritten due to inlining (all symbolic LCs are replaced by linear combinations of input/witness variables).

**Comment by weikengchen** [comment]
The implementation has been refactored in that the common elements between `inline_all_lcs` and `outline_lcs` are extracted.

@Pratyush @ValarDragon You may want to give a fresh reading to see if the logic flows right.

Note: The PR is not ready for merging, as I need to test it on Groth16 and Marlin.

**Comment by weikengchen** [comment]
And also, should we change `outline_lcs` to `outline_all_lcs`?

**Comment by weikengchen** [comment]
A bug has been fixed. It now passes on Groth16 and Marlin.

**Review by Pratyush — APPROVED** [review]
This PR looks great now, thanks for the effort =)

**Comment by weikengchen** [comment]
Opinions on changing `outline_lcs` to `outline_all_lcs`? This is a good chance to decide (since it has not been used anywhere else).

**Comment by Pratyush** [comment]
I think `outline_lcs` is better, as we're not outlining all the LCs, only those that are large, no?

**Comment by weikengchen** [comment]
That makes sense!

**Comment by ValarDragon** [comment]
This function is presumably not an external facing API right?

The external facing api should probably be "reduce_matrix_density"

**Comment by weikengchen** [comment]
Indeed it will be externally facing, in that Marlin calls outline_lcs instead of inline_all_lcs before converting the constraint system into a matrix.

Technically, outline_lcs is also mostly inlining.

**Comment by ValarDragon** [comment]
I see. I think we should make this function private, and make the external facing function `reduce_matrix_arity` (or `reduce_constraint_density` or something), which will then call this function. In the future, other matrix arity optimizations may make it in there.

**Review by ValarDragon — APPROVED** [review]
Looks great to me

**Comment by weikengchen** [comment]
I would call it `reduce_constraint_weight` since Ale uses `weight` in our paper on recursive Marlin.

**Comment by weikengchen** [comment]
Ok. The final change passes the CI.

I hereby need one person with write access to merge this PR.

**Comment by Pratyush** [comment]
In that case, should we provide a higher level API to inline_all_lcs too?

perhaps a single “finalize” method that takes in a rename OptimizationMode enum and then internally calls inline_lcs or outline_lcs? (This can be a separate PR)

**Pull request merged**
